### PR TITLE
Add the ability for fancy_ptycho to model detectors with spatially varying quantum efficiency

### PR DESCRIPTION
**Pull request by allevitan**
Recently I wound up working with a lot of data from detectors which have very... not flat flat fields. In the past I've seen this corrected by dividing out the flat-field, but this messes with the photon statistics a bit, and is especially bad for photon-counting data where suddenly you need to store integer-valued data as floating-point.

I think the right way to solve this is to introduce a layer into the forward model to simulate the variable detector quantum efficiency. So, in this PR I added that to fancy_ptycho. This comes along with:

* CDataset and subclasses like Ptycho2DDataset can now load and save a "quantum efficienct mask" qe_mask, basically a flat-field measurement. This isn't defined in the official cxi file format, but I think it was worth creating a way to do it because you'd want to be able to easily pass this around with the data
* When initializing FancyPtycho using from_dataset, now there is a new argument "use_qe_mask", which is set to "False" by default to preserve the original behavior. If set to "true", it will either load the quantum efficiency mask from the dataset, or (if no qe_mask is available) initialize the mask with ones.
* The qe_mask is defined as a torch.nn.Parameter, which (by default) has requires_grad set to False. I chose this because I could imagine there being value in refining the quantum efficiency mask in some cases, although some preliminary tests indicate that in general it's quite unstable to do. But, defining it as a parameter now makes it easy to test this in individual cases

I added a bit of test coverage, including using this model (with a qe_mask identically equal to 1) in one of the slow fancy_ptycho tests. I've also done a bit of testing with some datasets which are, unfortunately, not public.

I think it's possible that some bugs remain in the implementation of the qe_mask which might be surfaced later, although I've tried to test it rigorously. But I'd suggest that this can be ready to fold into master once we're confident that it isn't affecting any existing code that doesn't use this correction.

Let me know what you think!

### CODE REVIEW

**Review by yoshikisd — CHANGES_REQUESTED** [review]
Overall I think everything looks good. I ran a pytest on my machine and it passed.

There's a small typo in the documentation of __len__ in ptycho_2d_dataset.py that should be fixed. I've also added a suggestion in measurements.py that makes the code more compact.

**Comment by allevitan** [comment]
I folded in the comments from @yoshikisd.

There were two checks in test_analysis that had started sporadically failing - related to a tolerance that had been set at 1e-7 (but occasionally was returning a value slightly above that, like 1.03e-7, when run on the github runners). For example, one of the tests failed for this earlier merge to master: https://github.com/cdtools-developers/cdtools/commit/da695c7c06c55495bdda67dc66c475e3b0c427ff

I just raised the tolerance so tests should now consistently pass (there was no specific need for this tolerance to be exactly 1e-7, it was just checking that an initial setup was reasonable).

**Review by yoshikisd — APPROVED** [review]
These changes look good to me. Pytest also runs without faults on my machine. Let's merge this into the main branch.
